streams/src/main/java/org/apache/kafka/streams/processor/internals/PunctuateRatioSlidingWindow.java

```diff
@@ -0,0 +1,38 @@
+package org.apache.kafka.streams.processor.internals;
+
+import java.util.LinkedList;
+import java.util.Queue;
+
+public class PunctuateRatioSlidingWindow {
+    private final Queue<RatioTimeStamp> ratioQueue;
+    private final long windowSizeMillis;
+
+    public PunctuateRatioSlidingWindow(long windowSizeMillis) {
+        this.windowSizeMillis = windowSizeMillis;
+        this.ratioQueue = new LinkedList<>();
+    }
+
+    public void update(double ratio){
+        long currentTimeMillis = System.currentTimeMillis();
```

Comment:
We should never call `System.currentTimeMillis()` directly, but always use the `Time` interface. Otherwise we cannot mock it.